File: backend/app/schemas/campo.py
"""Schemas del área campo."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class TareaCreate(BaseModel):
    code: str = Field(min_length=1, max_length=20)
    title: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    priority: str = Field(default="medium", pattern="^(critical|high|medium|low)$")
    assigned_to: str | None = Field(default=None, max_length=60)
    lat: float | None = None
    lng: float | None = None


class TareaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    code: str
    title: str
    description: str | None
    priority: str
    estado: str
    assigned_to: str | None
    lat: float | None
    lng: float | None


class ReporteCreate(BaseModel):
    kind: str = Field(pattern="^(incident|voice|image|checkpoint|support)$")
    title: str = Field(min_length=1, max_length=160)
    body: str | None = Field(default=None, max_length=4000)
    lat: float | None = None
    lng: float | None = None
    transcribed: bool = False


class ReporteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    kind: str
    title: str
    body: str | None
    lat: float | None
    lng: float | None
    transcribed: bool
    created_by: str | None
    created_at: datetime


class EstadoUpdate(BaseModel):
    estado: str = Field(min_length=1, max_length=14)

File: frontend/src/types/campo.ts
import type { GeoPoint, ID, ISODateString } from "./common";

export type FieldUnitState =
  | "available"
  | "busy"
  | "en-route"
  | "support-requested"
  | "off-duty";

export const FIELD_STATE_LABEL: Record<FieldUnitState, string> = {
  available: "DISPONIBLE",
  busy: "OCUPADO",
  "en-route": "EN RUTA",
  "support-requested": "APOYO SOLICITADO",
  "off-duty": "FUERA SERVICIO",
};

export type FieldTaskStatus =
  | "incoming"     // recibida, sin aceptar
  | "accepted"     // aceptada por la unidad
  | "on-scene"     // llegada a destino
  | "completed"
  | "cancelled";

export type FieldTaskPriority = "critical" | "high" | "medium" | "low";

export interface FieldTask {
  id: ID;
  emergencyId: ID;
  code: string;            // T-204
  title: string;
  description: string;
  priority: FieldTaskPriority;
  status: FieldTaskStatus;
  assignedTo: string;      // call-sign
  location: GeoPoint;
  destination?: string;
  createdAt: ISODateString;
  updatedAt: ISODateString;
  acceptedAt?: ISODateString;
  completedAt?: ISODateString;
}

export type FieldReportKind =
  | "incident"
  | "voice"
  | "image"
  | "checkpoint"
  | "support";

export interface FieldReport {
  id: ID;
  emergencyId: ID;
  kind: FieldReportKind;
  title: string;
  body: string;
  location?: GeoPoint;
  imageUrl?: string;
  transcribed?: boolean;       // marca true cuando viene de voz
  createdAt: ISODateString;
  createdBy: string;
}

export interface FieldUnit {
  id: ID;
  callSign: string;            // BINOMIO-3
  operator: string;            // "Cabo Vega + Patrullero Soto"
  state: FieldUnitState;
  location: GeoPoint;
  battery: number;             // % radio/dispositivo
  lastPing: ISODateString;
  activeTaskId?: ID | null;
}

export interface NavigationTarget {
  id: ID;
  label: string;
  kind: "incident" | "shelter" | "rendezvous" | "checkpoint";
  location: GeoPoint;
  distanceKm?: number;         // computed when active
  bearing?: number;            // 0..360
}
